File: src/main/java/com/rephilo/luandun/leetcode/_101/Solution.java
package com.rephilo.luandun.leetcode._101;

import com.rephilo.luandun.leetcode.datastructure.TreeNode;

/**
 * 101
 *
 * @author rephilo
 */
public class Solution {
    /**
     * 和100题一致，只不过比较对象换一下
     *
     * @param root
     * @return
     */
    public boolean isSymmetric(TreeNode root) {
        return isMirror(root, root);
    }

    public boolean isMirror(TreeNode p, TreeNode q) {
        if (p == null && q == null) {
            return true;
        }
        if (p != null && q != null) {
            if (p.val == q.val) {
                return isMirror(p.left, q.right) && isMirror(p.right, q.left);
            }
        }

        return false;
    }
}

File: src/main/scala/com/rephilo/luandun/scala/leetcode/_101/Solution.scala
package com.rephilo.luandun.scala.leetcode._101

import com.rephilo.luandun.scala.leetcode.datastructure.TreeNode

/**
 * 101
 *
 * @author rephilo
 */
object Solution {
  /**
   *
   * @param root
   * @return
   */
  def isSymmetric(root: TreeNode): Boolean = {
    def isMirror(p: TreeNode, q: TreeNode): Boolean = {
      if (p == null && q == null) {
        return true
      }

      if (p != null && q != null) {
        if (p.value == q.value) {
          return isMirror(p.left, q.right) && isMirror(p.right, q.left)
        }
      }
      false
    }

    isMirror(root, root)
  }
}
